src/Streamly/Internal/FileSystem/Event/Windows.hs

```diff
@@ -0,0 +1,432 @@
+    . setModifiedAttribute s
+    . setModifiedSize s
+    . setModifiedLastWrite s
+    . setModifiedSecurity s    
+
+defaultConfig :: Config
+defaultConfig = setAllEvents On $ Config { watchRec = True, createFlags = 0 }
+
+getConfigFlag :: Config -> DWORD
+getConfigFlag Config{..} = createFlags
+
+getConfigRecMode :: Config -> BOOL
+getConfigRecMode Config{..} = watchRec
+
+data Event = Event
+    { eventFlags :: DWORD  
+    , eventRelPath :: String         
+    , eventRootPath :: String 
+    , totalBytes :: DWORD
+    } deriving (Show, Ord, Eq)  
+
+-- For reference documentation see:
+--    
+-- See https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-file_notify_information
+data FILE_NOTIFY_INFORMATION = FILE_NOTIFY_INFORMATION
+    { fniNextEntryOffset :: DWORD
+    , fniAction :: DWORD
+    , fniFileName :: String
+    } deriving Show    
+
+type LPOVERLAPPED_COMPLETION_ROUTINE = 
+    FunPtr ((DWORD, DWORD, LPOVERLAPPED) -> IO ())
+
+-- | A handle for a watch.
+getWatchHandle :: FilePath -> IO (HANDLE, FilePath)
+getWatchHandle dir = do
+    h <- createFile dir
+        fILE_LIST_DIRECTORY -- Access mode
+        (fILE_SHARE_READ .|. fILE_SHARE_WRITE) -- Share mode
+        Nothing -- security attributes
+        oPEN_EXISTING -- Create mode, we want to look at an existing directory
+        fILE_FLAG_BACKUP_SEMANTICS -- File attribute, nb NOT using OVERLAPPED since we work synchronously
+        Nothing -- No template file
+    return (h, dir)
+    
+-- For reference documentation see:
+--
+-- See https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-readdirectorychangesw
+-- Note that this API uses UTF-16 for file system paths: 
+-- 1. https://docs.microsoft.com/en-us/windows/win32/intl/unicode-in-the-windows-api
+-- 2. https://docs.microsoft.com/en-us/windows/win32/intl/unicode
+foreign import ccall safe 
+    "windows.h ReadDirectoryChangesW" c_ReadDirectoryChangesW 
+        :: HANDLE -> LPVOID -> DWORD -> BOOL -> DWORD -> LPDWORD 
+        -> LPOVERLAPPED -> LPOVERLAPPED_COMPLETION_ROUTINE -> IO BOOL
+
+readDirectoryChangesW 
+    :: HANDLE -> Ptr FILE_NOTIFY_INFORMATION -> DWORD 
+    -> BOOL -> FileNotificationFlag -> LPDWORD -> IO ()
+
+readDirectoryChangesW h buf bufSize wst f br =
+    failIfFalse_ "ReadDirectoryChangesW" $ c_ReadDirectoryChangesW h (castPtr buf) 
+    bufSize wst f br nullPtr nullFunPtr  
+
+readChangeEvents :: Ptr FILE_NOTIFY_INFORMATION -> String -> DWORD -> IO [Event]
+readChangeEvents pfni root bytesRet = do  
+    fni <- peekFNI pfni
+    let entry = Event{  eventFlags = fniAction fni 
+                      , eventRelPath = fniFileName fni  
+                      , eventRootPath = root
+                      , totalBytes = bytesRet
+                      }  
+        nioff = fromEnum $ fniNextEntryOffset fni   
+    entries <- if nioff == 0 
+                    then return [] 
+                    else readChangeEvents (pfni `plusPtr` nioff) root bytesRet
```

Comment:
The formatting of `Event` struct and `if` clause is missed. I gave examples about how to format it in the previous comments.